Adding custom routes to the mux. Cannot add root route. ("/")

Hello, I have feedback on slack. I tested in my app and I think this is a bug

> https://grpc-ecosystem.github.io/grpc-gateway/docs/operations/inject_router/ 

Follow this example, I want to add some routes of my own.

> err := mux.HandlePath("GET", "/abc", xxxx)
> fmt.Println(err)

This is normal.  err == nil, I can successfully access through the browser www.my-url.com/abc

But!!!!!!
> err := mux.HandlePath("GET", "/", xxxx)
> fmt.Println(err)

This is abnormal.   I cannot access through www.my-url.com/, and the `err` return
```
parsing path pattern: segment neither wildcards, literal or variable: expected "{" but got "\x00": /
```

This is an example. Edit by https://grpc-ecosystem.github.io/grpc-gateway/docs/operations/inject_router/.

This example will panic. and show error message.


main.go:

```go
package main

import (
	"context"
	"net/http"
	"testgw/helloworld"

	"github.com/grpc-ecosystem/grpc-gateway/v2/runtime"
)

func main() {
	ctx := context.TODO()
	mux := runtime.NewServeMux()
	// Register generated routes to mux
	err := helloworld.RegisterGreeterHandlerServer(ctx, mux, &GreeterServer{})
	if err != nil {
		panic(err)
	}
	// Register custom route for  GET /hello/{name}
	// This route is OK!
	err = mux.HandlePath("GET", "/hello/{name}", func(w http.ResponseWriter, r *http.Request, pathParams map[string]string) {
		w.Write([]byte("hello " + pathParams["name"]))
	})
	if err != nil {
		panic(err)
	}
	// This route is OK!
	err = mux.HandlePath("GET", "/test/{name}", func(w http.ResponseWriter, r *http.Request, pathParams map[string]string) {
		w.Write([]byte("test with name " + pathParams["name"]))
	})
	if err != nil {
		panic(err)
	}
	// This route is OK!
	err = mux.HandlePath("GET", "/test", func(w http.ResponseWriter, r *http.Request, pathParams map[string]string) {
		w.Write([]byte("this is test " + pathParams["name"]))
	})
	if err != nil {
		panic(err)
	}
	// This route will be PANIC
	// Return: panic: parsing path pattern: segment neither wildcards, literal or variable: expected "{" but got "\x00": /
	err = mux.HandlePath("GET", "/", func(w http.ResponseWriter, r *http.Request, pathParams map[string]string) {
		w.Write([]byte("this is homepage"))
	})
	if err != nil {
		panic(err)
	}
	http.ListenAndServe(":8080", mux)
}

// GreeterServer is the server API for Greeter service.
type GreeterServer struct {

}

// SayHello implement to say hello
func (h *GreeterServer) SayHello(ctx context.Context, req *helloworld.HelloRequest) (*helloworld.HelloReply, error) {
	return &helloworld.HelloReply{
		Message: "hello " + req.Name,
	}, nil
}
```


return
```shell
panic: parsing path pattern: segment neither wildcards, literal or variable: expected "{" but got "\x00": /

goroutine 1 [running]:
main.main()
        C:/Users/User/go/src/testgw/main.go:46 +0x2c4
exit status 2
```



# How to run example:
1.mkdir testgw in GOPATH & go mod init & go mod vendor
2.Copy MY Example main.go code into your project
3.Copy  grpc-gateway/examples/internal/helloworld/ this directory into your project

--- a/internal/httprule/compile.go
+++ b/internal/httprule/compile.go
@@ -96,6 +96,10 @@ func (t template) Compile() Template {
 		if op.str == "" {
 			ops = append(ops, op.num)
 		} else {
+			// eof segment literal represents the "/" path pattern
+			if op.str == eof {
+				op.str = ""
+			}
 			if _, ok := consts[op.str]; !ok {
 				consts[op.str] = len(pool)
 				pool = append(pool, op.str)

--- a/internal/httprule/compile_test.go
+++ b/internal/httprule/compile_test.go
@@ -23,6 +23,13 @@ func TestCompile(t *testing.T) {
 		{},
 		{
 			segs: []segment{
+				literal(eof),
+			},
+			ops:  []int{int(utilities.OpLitPush), 0},
+			pool: []string{""},
+		},
+		{
+			segs: []segment{
 				wildcard{},
 			},
 			ops: []int{int(utilities.OpPush), operandFiller},

--- a/internal/httprule/parse.go
+++ b/internal/httprule/parse.go
@@ -118,6 +118,10 @@ type parser struct {
 
 // topLevelSegments is the target of this parser.
 func (p *parser) topLevelSegments() ([]segment, error) {
+	if _, err := p.accept(typeEOF); err == nil {
+		p.tokens = p.tokens[:0]
+		return []segment{literal(eof)}, nil
+	}
 	segs, err := p.segments()
 	if err != nil {
 		return nil, err

--- a/internal/httprule/parse_test.go
+++ b/internal/httprule/parse_test.go
@@ -141,6 +141,20 @@ func TestParseSegments(t *testing.T) {
 		want   []segment
 	}{
 		{
+			tokens: []string{eof},
+			want: []segment{
+				literal(eof),
+			},
+		},
+		{
+			// Note: this case will never arise as tokenize() will never return such a sequence of tokens
+			// and even if it does it will be treated as [eof]
+			tokens: []string{eof, "v1", eof},
+			want: []segment{
+				literal(eof),
+			},
+		},
+		{
 			tokens: []string{"v1", eof},
 			want: []segment{
 				literal("v1"),
@@ -307,10 +321,6 @@ func TestParseSegmentsWithErrors(t *testing.T) {
 		{
 			// invalid percent-encoding
 			tokens: []string{"a%2z", eof},
-		},
-		{
-			// empty segments
-			tokens: []string{eof},
 		},
 		{
 			// unterminated variable

--- a/runtime/mux_test.go
+++ b/runtime/mux_test.go
@@ -419,6 +419,12 @@ var defaultRouteMatcherTests = []struct {
 	valid  bool
 }{
 	{
+		"Test route /",
+		"GET",
+		"/",
+		true,
+	},
+	{
 		"Simple Endpoint",
 		"GET",
 		"/v1/{bucket}/do:action",

